server/webapp/WEB-INF/rails.new/app/controllers/admin/materials/pluggable_scm_controller.rb

```diff
@@ -0,0 +1,212 @@
+        include ::ConfigUpdate::LoadConfig
+
+        def initialize params, user, security_service, material
+          super(params, user, security_service)
+          @material = material
+        end
+
+        def node(cruise_config)
+          cruise_config
+        end
+
+        def subject(cruise_config)
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).materialConfigs().last()
+        end
+
+        def update(cruise_config)
+          scm = cruise_config.getSCMs().find(params[:material][:scmId])
+          @material.setSCMConfig(scm)
+
+          @material.setConfigAttributes(params[:material])
+
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).addMaterialConfig(@material)
+        end
+      end.new(params, current_user.getUsername(), security_service, @material)
+    end
+
+    def get_create_command
+      Class.new(::ConfigUpdate::SaveAsPipelineAdmin) do
+        include ::ConfigUpdate::LoadConfig
+
+        def initialize params, user, security_service, material, pluggable_scm_service
+          super(params, user, security_service)
+          @material = material
+          @pluggable_scm_service = pluggable_scm_service
+        end
+
+        def node(cruise_config)
+          cruise_config
+        end
+
+        def subject(cruise_config)
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).materialConfigs().last()
+        end
+
+        def update(cruise_config)
+          scm = @material.getSCMConfig()
+          scm.setConfigAttributes(params[:material])
+          scm.ensureIdExists
+          params[:material][PluggableSCMMaterialConfig::SCM_ID] = scm.getSCMId
+
+          @pluggable_scm_service.validate(scm)
+
+          @material.setConfigAttributes(params[:material])
+
+          cruise_config.getSCMs().add(scm)
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).addMaterialConfig(@material)
+        end
+      end.new(params, current_user.getUsername(), security_service, @material, pluggable_scm_service)
+    end
+
+    def get_update_command
+      Class.new(::ConfigUpdate::SaveAsPipelineAdmin) do
+        include ::ConfigUpdate::LoadConfig
+
+        def initialize params, user, security_service, material, index, pluggable_scm_service
+          super(params, user, security_service)
+          @material = material
+          @index = index
+          @pluggable_scm_service = pluggable_scm_service
+        end
+
+        def node(cruise_config)
+          cruise_config
+        end
+
+        def subject(cruise_config)
+          pipeline = cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name]))
+          #Since the fingerprint that got submitted could be different from the new fingerprint after the setConfigAttributes, use the index.
+          if @index
```

Comment:
Why will @index be nil? What is the consequence of not returning material_config, in that case?